### ISSUE TITLE
frama-clang failed to install

### ISSUE THREAD

**Issue by Runingtime** [Frama-C/Frama-C-snapshot#10]
Hi,
    I'm trying to install the Frama-Clang plugin following the [instructions on the website](https://frama-c.com/frama-clang.html). But the command 

> ./configure

failed to execute and the messages are shown below:

```
checking for frama-c-gui... yes
checking for Makefile.config.in... yes
frama_clang... yes
checking for g++... g++
checking whether the C++ compiler works... yes
checking for C++ compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C++ compiler... yes
checking whether g++ accepts -g... yes
checking for camlp4o... yes
checking for clang... clang
checking for clang++... clang++
checking for llvm-config... llvm-config
checking for /usr/lib/llvm-4.0/include/clang... no
configure: WARNING: frama_clang disabled because clang dev headers not found.
configure: frama_clang: no
configure: creating ./config.status
config.status: creating ./Makefile.config
```

I don't know how to solve this problem, any suggestions?

**Comment by maroneze** [comment]
Frama-Clang's `configure` is not finding `clang`'s headers in the expected directory. Could you please provide what system are you using (if Linux, which distribution, otherwise if it's a macOS or some other system)? Also, it may be useful to know which versions of related libraries (libclang-dev, llvm-config, etc.) are installed in your system.

There are several possible causes:
- clang/libclang are not installed;
- different versions of llvm, llvm-config, clang and/or libclang are installed;
- your system places headers using a different structure (apparently `nix` and a few other systems do that).

Basically, what frama-clang does is run `llvm-config --includedir` to find out where `clang`'s headers are, and it's not finding them there. With the above information we may be able to help further.

**Comment by maroneze** [comment]
The Ubuntu 16.04 (xenial) package seems to contain Frama-C Magnesium (20151002). I cannot reproduce the exact error message you are having, but that version of Frama-C is somewhat old and not compatible with frama-clang-0.0.6 (the AST has changed since then).

I'm afraid you'll have to try installing Frama-C from opam. Which means you might run into package conflicts if you mix OCaml packages from Ubuntu and those from opam.

Here's what we usually recommend in such cases:

1. Uninstalling Frama-C from apt-get (`apt-get remove frama-c`);
2. Ideally, uninstalling most `ocaml-*` packages from apt-get as well, to ensure they don't mix up with those which will be installed by opam;
3. Installing opam (`apt-get install opam`);
4. Initializing opam (`opam init`), without forgetting to set the environment at the end, as written in its instructions (`eval $(opam config env)`);
5. Installing Frama-C via opam (`opam install frama-c`), which should get version 17 (Chlorine-20180502);
6. Compiling Frama-Clang using that version of Frama-C.

The last step should work since you already have the C libraries required, however uninstalling the ocaml Ubuntu packages might make apt-get uninstall these as well, so you might need to do something such as:

- `opam install depext`
- `opam depext frama-c`

These commands should help installing the Ubuntu libraries required by Frama-C (GMP, GTK, etc), in case a `opam install frama-c` fails.

Using opam is not as straightforward as using apt-get, but it offers more up-to-date packages, and in the case of this release of Frama-Clang it will be necessary, since it is very recent.

**Comment by Runingtime** [comment]
@maroneze thank you for the detailed instructions! 
I compiled frama-c from source a few hours ago and it works then. 
Can you further tell me how to parse simple c++ programs with it? Here is the question I asked before:
[how to use frama-clang](https://stackoverflow.com/questions/51586499/how-to-use-frama-clang-to-parse-c-programs)

**Comment by maroneze** [comment]
I'm not a user of Frama-Clang, so I'm afraid I won't be able to give a precise answer to your question, however it is likely that the usage of STL is an issue there.

While the example *looks* simple from the point of view of a programmer, a "Hello World" in C++ expands to a huge amount of code. For instance, if I run `g++ -E hello.cpp`, I get >28000 lines of preprocessed output; with `clang` I "only" get ~18000, but it's still gigantic.

I'd recommend trying an example without the STL. For instance, if I run `frama-c test.cpp` (with Frama-Clang installed) on the code below, it parses without issues. I can even add `-val` to run the Eva plug-in afterwards.

```
class Rectangle {
  int width, height;
public:
  void set_values (int,int);
  int area() {return width*height;}
};

void Rectangle::set_values (int x, int y) {
  width = x;
  height = y;
}

int main () {
  Rectangle rect;
  rect.set_values (3,4);
  return rect.area();
}
```

From the point of view of embedded code, not using the STL is not that uncommon, and as far as I know that was the initial use case for Frama-Clang.

**Comment by Runingtime** [comment]
All right. 
Thank you again for everything you've done.
